### PR TITLE
[naga] parse and validate `@must_use` attribute

### LINKED ISSUE #5186: Support `@must_use` function attribute in WGSL
We do not currently support the `@must_use` function attribute, despite being specified in the current v1 spec. draft. This was introduced in early 2023; see [`gpuweb/gpuweb`#3862](https://github.com/gpuweb/gpuweb/pull/3862), [`gpuweb/gpuweb`#3819](https://github.com/gpuweb/gpuweb/issues/3819).

# Repro steps
<sup>Ideally, a runnable example we can check out.</sup>

1. With this `in.wgsl` input file:

	```wgsl
	@must_use 
	fn helper() -> u32 {
		return 2;
	}
	
	@compute @workgroup_size(1)
	fn comp_main() {
		let asdf = helper();
	}
	```
	
	…run this on latest `trunk` (ATOW, d239361c1f5d5cfb8cee73382b1e8bc8476e3b29):

    ```
    $ cargo run -p naga-cli --all-features -- in.wgsl
    ```

# Expected vs observed behavior
<sup>Clearly describe what you get, and how it goes across your expectations.</sup>

This should compile without error. The `asdf` binding is unused, but that should, at most, be a warning. Instead, the current output of repro steps is:

```
Could not parse WGSL:
error: unknown attribute: 'must_use'
  ┌─ in.wgsl:1:2
  │
1 │ @must_use
  │  ^^^^^^^^ unknown attribute
```

### PR DESCRIPTION
**Pull request by turbocrime**
# ready for review

~~to be revised now that i can run conformance tests~~

not worrying about conformance tests, as cts_runner seems disused.

----

**Connections**

resolves https://github.com/gfx-rs/wgpu/issues/5186

**Description**

When the wgsl parser encounters the `@must_use` function attribute, it should not error.

This PR will recognize the attribute and emit a new field in function AST. Both the parser and the validator stage will use this attribute.

The parser requires that the annotated function should return a value, and requires that the function call appears in an expression.

The validator requires that return value is referenced or assigned.

A few tests are added covering these behaviors and more.

**Testing**
- tested with wgpu test suite
- tested by building Firefox nightly with this change

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

**Comment by turbocrime** [comment]
Looking at discussion in the `gpuweb` repository linked in the issue, it seems like

- wgsl builtins may have the attribute `must_use`
- a wgsl compiler should reject bare calls to `must_use`
- `must_use` validation should happen at compile-time and not at run-time

I expect that many particularities may complicate validation, so tests are provided for slightly complex valid examples.

I'm unfamiliar with how naga's AST behaves, and looking through the validation code, it didn't seem obvious where/how to validate that a `CallResult` is 'used', and what qualifies as 'used'.

**Comment by turbocrime** [comment]
~~Are there likely to be important validation differences between constant and variable values?~~

File: naga/src/front/wgsl/parse/ast.rs [hunk 1]
```diff
@@ -114,6 +114,7 @@ pub struct FunctionArgument<'a> {
 pub struct FunctionResult<'a> {
     pub ty: Handle<Type<'a>>,
     pub binding: Option<Binding<'a>>,
+    pub must_use: bool,
 }
```
Comment (turbocrime) [inline naga/src/front/wgsl/parse/ast.rs:118]: `must_use` is a required member of wgsl ast `FunctionResult`.

File: naga/src/front/wgsl/parse/mod.rs [hunk 2]
```diff
@@ -2382,7 +2383,11 @@ impl Parser {
         let result = if lexer.skip(Token::Arrow) {
             let binding = self.varying_binding(lexer, &mut ctx)?;
             let ty = self.type_decl(lexer, &mut ctx)?;
-            Some(ast::FunctionResult { ty, binding })
+            Some(ast::FunctionResult {
+                ty,
+                binding,
+                must_use,
+            })
```
Comment (turbocrime) [inline naga/src/front/wgsl/parse/mod.rs:2392]: `ast::FunctionResult` creations are updated

File: naga/src/front/wgsl/parse/mod.rs [hunk 3]
```diff
@@ -2455,6 +2460,7 @@ impl Parser {
         let (mut bind_index, mut bind_group) =
             (ParsedAttribute::default(), ParsedAttribute::default());
         let mut id = ParsedAttribute::default();
+        let mut must_use = ParsedAttribute::default();
```
Comment (turbocrime) [inline naga/src/front/wgsl/parse/mod.rs]: attribute is handled in `global_decl`

is there another location in which `@must_use` might validly appear?

should invalid locations recognize the attribute at all?

possibly, some global declarations should not have the `@must_use` attribute. no checks are performed.

File: naga/src/front/wgsl/parse/mod.rs [hunk 4]
```diff
@@ -2333,6 +2333,7 @@ impl Parser {
         diagnostic_filter_leaf: Option<Handle<DiagnosticFilterNode>>,
         out: &mut ast::TranslationUnit<'a>,
         dependencies: &mut FastIndexSet<ast::Dependency<'a>>,
+        must_use: bool,
```
Comment (turbocrime) [inline naga/src/front/wgsl/parse/mod.rs]: i added a parameter to `function_decl` for this attribute.

File: naga/tests/in/must-use-questionable.wgsl
Comment (turbocrime) [inline naga/tests/in/must-use-questionable.wgsl:1]: these examples seem to be permitted, but as a rule never result in use of the return at runtime.

**Review by turbocrime — COMMENTED** [review]
some commentary

**Comment by turbocrime** [comment]
Experimented with validation over the weekend. It seems like AST cannot distinguish between the two examples below, one of which was used in the issue as a 'should pass' example, one of which is a clear failure.

```wgsl
// an unused call (should fail)
fn actually_unused() {
	helper();
}

// a used call (should pass)
// but `asdf` is unused
@compute @workgroup_size(1)
fn assignment_unused() {
	let asdf = helper();
}
```

**Comment by turbocrime** [comment]
i'm going to revise this now that i can run conformance tests after #6840 

it is not presently ready for review

**Comment by sagudev** [comment]
> i'm going to revise this now that i can run conformance tests after #6840

Unfortunately am not sure how relevant information from cts_runner actually is, as there is no concept of (bad) expectations (so only selected tests are run) and it uses old CTS version. The most useful results come from browsers running CTS, either Firefox or Servo as both run all test and have expectations properly set, so you know what results become different using your PR.

In servo you can simply override [wgpu in Cargo.toml](https://github.com/sagudev/servo/commit/aa3b33e1b025af28f2830493601b805b01a5b2e6) (ensuring that your PR bases from same revision of wgpu previous used in servo) then run `./mach try webgpu` in servo folder to trigger CI run of CTS in GitHub Actions. I did initial CTS run for this PR here: https://github.com/sagudev/servo/actions/runs/12568503599/attempts/1#summary-35037584122 (note that there are some flaky crashes in webgpu:shader,execution,expression due to cyclic JS imports, so you can ignore them).

**Comment by turbocrime** [comment]
yes, realizing this since i patched things together. i was hoping to have canonical tests for edge cases and ambiguities, that could execute in a nice automated way, but it seems like the validation tests can't run in cts_runner without more api work.

maybe this repo should remove cts_runner and the documentation that suggests using it, if it's no longer relevant.

thanks for the instructions for executing in servo's ci! this is very useful

**Comment by jamienicol** [comment]
I've ran the CTS without and with this in firefox and it looks good. There are no new failures, and there are a few new passes as expected. There are also quite a few more places where previously we failed due to `parsing error: unknown attribute: must_use`, but with that fixed now we fail for another reason. (Seems to be type conversion mostly AFAICS)

As you noted above this does have the limitation that it doesn't ensure the result of `@must_use` _builtin_ functions are actually used. But let's not let perfect be the enemy of good. We can save that for a follow-up. I think this is mostly ready to go.

@turbocrime do you still have changes planned for this, or would you like me to review as-is? Or if you're too busy I'm happy to take this over the line for you. LMK!

File: naga/tests/wgsl_errors.rs [hunk 5]
```diff
@@ -1480,6 +1480,35 @@ fn dead_code() {
...
+                source: naga::valid::FunctionError::InvalidCall {
+                    function: _callee,
+                    error: naga::valid::CallError::ResultNotUsed(_result),
+                },
+            },
+        )
+        if caller_name == "useless"
+    }
+}
+
```
Comment (turbocrime) [inline naga/tests/wgsl_errors.rs]: not sure how to test validation if the parser will reject the shader. perhaps i could manually create the AST by parsing a shader without the attribute and then modifying it.

**Comment by turbocrime** [comment]
> As you noted above this does have the limitation that it doesn't ensure the result of `@must_use` _builtin_ functions are actually used. But let's not let perfect be the enemy of good. We can save that for a follow-up. I think this is mostly ready to go.

i assume the builtins aren't presently 'annotated'. if must_use builtins are expressed correctly in AST, perhaps validation phase can catch that.

validation code is present and works if i remove the parser error, but i'm not sure how to test it when the parser throws. i'm not sure if the validation code is actually useful - would the parser know about builtins' attributes at parsing time? will the parsing step ever be skipped? do we care about that validation if the parsing step is skipped?

> @turbocrime do you still have changes planned for this, or would you like me to review as-is? Or if you're too busy I'm happy to take this over the line for you. LMK!

i had a local commit sitting around that reverted some random junk. pushed that up, ready for you to take a look.

i will have way less free time starting monday so feel free to take over.

**Comment by turbocrime** [comment]
if you'd like to see the validation phase successfully detect an unused call, see f028d07c1d444a9795cd26ec87208f8842856407 where i have disabled the parser failures

**Comment by turbocrime** [comment]
this is covered by the parser, but perhaps the validator should also error if it encounters a call to some `@must_use` that somehow returns nothing, whether it's in an expression or a statement. i experimented briefly but couldn't get the parser to emit an example.

File: naga/src/front/wgsl/error.rs [hunk 6]
```diff
@@ -262,6 +262,8 @@ pub(crate) enum Error<'a> {
         found: u32,
     },
     FunctionReturnsVoid(Span),
+    FunctionMustUseUnused(Span),
+    FunctionMustUseNothing(Span, Span),
```
Comment (jamienicol) [inline naga/src/front/wgsl/error.rs]: Let's rename this `FunctionMustUseReturnsVoid` to be consistent with the `FunctionReturnsVoid` error just above.

File: naga/src/front/wgsl/parse/mod.rs [hunk 7]
```diff
@@ -2383,7 +2384,17 @@ impl Parser {
...
+            let must_use = must_use.is_some();
+            Some(ast::FunctionResult {
+                ty,
+                binding,
+                must_use,
+            })
+        } else if let Some(must_use) = must_use {
+            return Err(Error::FunctionMustUseNothing(
+                must_use,
+                self.pop_rule_span(lexer).until(&lexer.next().1),
```
Comment (jamienicol) [inline naga/src/front/wgsl/parse/mod.rs]: this can be replaced with `self.peek_rule_span(lexer)`.

File: naga/src/valid/function.rs [hunk 8]
```diff
@@ -22,6 +22,8 @@ pub enum CallError {
     ResultAlreadyInScope(Handle<crate::Expression>),
     #[error("Result expression {0:?} is populated by multiple `Call` statements")]
     ResultAlreadyPopulated(Handle<crate::Expression>),
+    #[error("Result expression {0:?} is not used")]
```
Comment (jamienicol) [inline naga/src/valid/function.rs]: I think we should just remove this error and the check during validation altogether (and therefore the test too). It's already caught during parsing, and with a nicer error message.

When we add support for builtin functions, perhaps that will be caught by the validator rather than the parser. _If_ that's the case then we can easily add this back. But no point having it here currently doing nothing.

File: naga/tests/in/must-use.wgsl [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+fn use_assign_var() -> i32 {
+    var q = use_me();
+    return q;
+}
+
+fn use_assign_let() -> i32 {
+    let q = use_me();
+    return q;
+}
```
Comment (jamienicol) [inline naga/tests/in/must-use.wgsl:19]: It'd be nice to have a [phony assignment](https://www.w3.org/TR/WGSL/#phony-assignment) check too, since that's how the spec specifically suggests to work avoid must_use errors.

eg `_ = use_me();`

File: naga/tests/wgsl_errors.rs (hunk 5 above)
Comment (jamienicol) [reply naga/tests/wgsl_errors.rs]: As I said higher up, let's just remove the validation error and this test too

File: naga/src/front/wgsl/lower/mod.rs [hunk 10]
```diff
@@ -2208,6 +2209,15 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                     .collect::<Result<Vec<_>, _>>()?;
 
                 let has_result = ctx.module.functions[function].result.is_some();
+                let must_use = ctx.module.functions[function]
+                    .result
+                    .as_ref()
+                    .is_some_and(|result| result.must_use);
+
+                if must_use && is_statement {
+                    return Err(Error::FunctionMustUseUnused(span));
```
Comment (jamienicol) [inline naga/src/front/wgsl/lower/mod.rs]: `function.span` (where `function` is the argument to `call()`) is a better span to use than `span`. However, the argument `function` is being shadowed by this pattern match arm on line 2187. We can either rename `function` in the match arm, or create a new variable `function_span` prior to the match block.

This additionally means you don't have to `.split_once('(')` the span in error.rs

File: naga/src/front/wgsl/error.rs [hunk 11]
```diff
@@ -820,6 +822,31 @@ impl<'a> Error<'a> {
...
+                message: "must use result".into(),
+                labels: vec![(call, "result unused".into())],
+                notes: vec![
+                    format!(
+                        "function '{}' is declared with `@must_use` attribute",
+                        &source[call].split_once('(').unwrap().0
+                    ),
+                    "call the function as part of an expression".into(),
+                ],
+            },
```
Comment (jamienicol) [inline naga/src/front/wgsl/error.rs]: I think we can make these error messages a bit better.

- Let's make the `message` more descriptive, eg "unused return value from function annotated with @must_use"
- The label with the span is good but I don't think it needs an additional description as it would just be repeating the line above, so use an empty string there.
- The notes are good, but I think we can improve the second one further by taking the suggested fix from the spec: "use a phony assignment or declare a value using the function call as the initializer."

File: naga/src/front/wgsl/error.rs [hunk 12]
```diff
@@ -820,6 +822,31 @@ impl<'a> Error<'a> {
...
+                labels: vec![
+                    (attr, "requires result".into()),
+                    (signature, "no result".into()),
+                ],
+                notes: vec![
+                    format!(
+                        "function '{}' is declared with `@must_use` attribute",
+                        &source[signature].split_once('(').unwrap().0
+                    ),
+                    "declare a return type or remove the attribute".into(),
```
Comment (jamienicol) [inline naga/src/front/wgsl/error.rs]: And for this one:
- message could be eg: "function annotated with @must_use but does not return any value"
- again I don't think the text labels actually add anything, just highlighting the spans is good enough, so would use empty strings for the labels
- I don't think the first note adds anything as the message and label already draw attention to that. the second note is great

**Review by jamienicol — COMMENTED** [review]
This is great just a few minor changes to make I think. Happy to take those on myself as you said you're busier from today. Thanks!

**Comment by jamienicol** [comment]
Oh and one more thing missing, we need to add the new "must-use.wgsl" file to snapshots.rs so that we take snapshots of it during a test run. I think given none of the backends support this we're only really interested in a) whether the shader parses/validates and b) whether the validation has the must_use flag. So we can only set the `Targets::IR` flag.

If/when we implement support for this in wgsl-out we can enable the Targets::WGSL flag too

**Comment by jamienicol** [comment]
I've rebased and pushed another commit with the changes I mentioned above, and also added you to the changelog @turbocrime

File: naga/src/front/wgsl/lower/mod.rs [hunk 13]
```diff
@@ -2208,6 +2210,15 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                     .collect::<Result<Vec<_>, _>>()?;
 
                 let has_result = ctx.module.functions[function].result.is_some();
+                let must_use = ctx.module.functions[function]
+                    .result
+                    .as_ref()
+                    .is_some_and(|result| result.must_use);
```
Comment (teoxoy) [inline naga/src/front/wgsl/lower/mod.rs]: Since all the validation happens in the frontend it doesn't seem necessary to have `FunctionResult.must_use` as part of the IR.

Would it work to add the `must_use` bool to `LoweredGlobalDecl::Function` instead?

**Comment by teoxoy** [comment]
Left a comment, this looks good otherwise!

File: naga/src/front/wgsl/lower/mod.rs (hunk 13 above)
Comment (jamienicol) [reply naga/src/front/wgsl/lower/mod.rs]: Sounds good. I've made that change, and reverted the other related changes, ie the helper functions FunctionResult::new() and FunctionResult::bound(), and the IR snapshots

**Review by teoxoy — APPROVED** [review]

**Comment by turbocrime** [comment]
thanks for the improvements! i completely agree.

happy to see this merge :)
